controlplane/kubeadm/internal/control_plane.go

```diff
@@ -0,0 +1,204 @@
+			machinefilters.Not(machinefilters.MatchesConfigurationHash(hash.Compute(&c.KCP.Spec))),
+		)
+	}
+	return requireUpgrade
+}
+
+// FailureDomainWithMost returns the failure domain with the most number of machines.
+// Used when scaling down.
+func (c *ControlPlane) FailureDomainWithMost() *string {
+	// See if there are any Machines that are not in currently defined failure domains first.
+	notInFailureDomains := c.Machines.Filter(
+		machinefilters.Not(machinefilters.InFailureDomains(c.Cluster.Status.FailureDomains.FilterControlPlane().GetIDs()...)),
+	)
+	if len(notInFailureDomains) > 0 {
+		// return the failure domain for the oldest Machine not in the current list of failure domains
+		// this could be either nil (no failure domain defined) or a failure domain that is no longer defined
+		// in the cluster status.
+		return notInFailureDomains.Oldest().Spec.FailureDomain
+	}
+
+	// Otherwise pick the currently known failure domain with the most Machines
+	return PickMost(c.Cluster.Status.FailureDomains.FilterControlPlane(), c.Machines)
+}
+
+// FailureDomainWithFewest returns the failure domain with the fewest number of machines.
+// Used when scaling up.
+func (c *ControlPlane) FailureDomainWithFewest() *string {
+	if len(c.Cluster.Status.FailureDomains.FilterControlPlane()) == 0 {
+		return nil
+	}
+	return PickFewest(c.Cluster.Status.FailureDomains.FilterControlPlane(), c.Machines)
+}
+
+// InitialControlPlaneConfig returns a new KubeadmConfigSpec that is to be used for an initializing control plane.
+func (c *ControlPlane) InitialControlPlaneConfig() *bootstrapv1.KubeadmConfigSpec {
+	bootstrapSpec := c.KCP.Spec.KubeadmConfigSpec.DeepCopy()
+	bootstrapSpec.JoinConfiguration = nil
+	return bootstrapSpec
+}
+
+// JoinControlPlaneConfig returns a new KubeadmConfigSpec that is to be used for joining control planes.
+func (c *ControlPlane) JoinControlPlaneConfig() *bootstrapv1.KubeadmConfigSpec {
+	bootstrapSpec := c.KCP.Spec.KubeadmConfigSpec.DeepCopy()
+	bootstrapSpec.InitConfiguration = nil
+	bootstrapSpec.ClusterConfiguration = nil
+	return bootstrapSpec
+}
+
+// GenerateKubeadmConfig generates a new kubeadm config for creating new control plane nodes.
+func (c *ControlPlane) GenerateKubeadmConfig(spec *bootstrapv1.KubeadmConfigSpec) *bootstrapv1.KubeadmConfig {
+	// Create an owner reference without a controller reference because the owning controller is the machine controller
+	owner := metav1.OwnerReference{
+		APIVersion: controlplanev1.GroupVersion.String(),
+		Kind:       "KubeadmControlPlane",
+		Name:       c.KCP.Name,
+		UID:        c.KCP.UID,
+	}
+
+	bootstrapConfig := &bootstrapv1.KubeadmConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            names.SimpleNameGenerator.GenerateName(c.KCP.Name + "-"),
+			Namespace:       c.KCP.Namespace,
+			Labels:          ControlPlaneLabelsForClusterWithHash(c.Cluster.Name, c.ConfigurationHash()),
+			OwnerReferences: []metav1.OwnerReference{owner},
```

Comment:
Do we want to make this a controller owner ref?

Reply:
the reconcile is not currently making this a controller owner reference, which is where this code originated. Do you think this should be a controller owner ref?